OpenKeychain/src/main/java/org/sufficientlysecure/keychain/keyimport/WebKeyDirectoryClient.java

```diff
@@ -0,0 +1,203 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import timber.log.Timber;
+
+
+public class WebKeyDirectoryClient implements KeyserverClient {
+
+    public static WebKeyDirectoryClient getInstance() {
+        return new WebKeyDirectoryClient();
+    }
+
+    private WebKeyDirectoryClient() { }
+
+    private static Pattern EMAIL_PATTERN = Pattern.compile("^\\s*(.+)@(.+)\\s*$");
+
+    @Override
+    public List<ImportKeysListEntry> search(String name, ParcelableProxy proxy)
+            throws QueryFailedException, QueryNeedsRepairException {
+        Timber.d("Web Key Directory import: " + name + " using Proxy: " + proxy.getProxy());
+        List<ImportKeysListEntry> entry = new ArrayList<>(1);
+
+        byte[] data = query(name, proxy);
+        // if we're here that means key retrieval succeeded,
+        // would have thrown an exception otherwise
+        try {
+            try {
+                entry.add(getEntry(name, data));
+            } catch (UnsupportedOperationException e) {
+                Timber.e("Parsing retrieved Web Key Directory key failed!");
+            }
+        } catch (PgpGeneralException | IOException e) {
+            Timber.e(e, "Failed parsing key from Web Key Directory during search");
+            throw new QueryFailedException("No valid key found on Web Key Directory");
+        }
+        return entry;
+    }
+
+    @Override
+    public String get(String name, ParcelableProxy proxy) throws QueryFailedException {
+        throw new UnsupportedOperationException("Returning armored key from Web Key Directory not supported");
+    }
+
+    private byte[] query(String user, ParcelableProxy proxy) throws QueryFailedException {
+        try {
+            Matcher matcher = EMAIL_PATTERN.matcher(user);
+
+            if (!matcher.matches()) {
+                throw new QueryFailedException("could not look up Web Key Directory for non-email input: " + user);
+            }
+
+            String domain = matcher.group(2);
+            String localPart = matcher.group(1);
+            String encodedPart = encodeZBase32(toSHA1(localPart.getBytes()));
+
+            URL url = new URL("https://" + domain + "/.well-known/openpgpkey/hu/" + encodedPart);
+
+            Timber.d("fetching from Web Key Directory with: " + url + " proxy: " + proxy.getProxy());
+
+            Request request = new Request.Builder()
+                    .url(url)
+                    .build();
+
+            OkHttpClient client = OkHttpClientFactory.getClientPinnedIfAvailable(url, proxy.getProxy());
+            Response response = client.newCall(request).execute();
+
+            if (response.isSuccessful()) {
+                return response.body().bytes();
+            } else {
+                // probably a 404 indicating that the key does not exist
```

Comment:
maybe check for a 5xx type error and return a different kind of message (server overloaded)